jwst/model_blender/blendrules.py

```diff
@@ -81,6 +83,29 @@ def last(items):
         return items[-1]
     return None
 
+def mindate(items):
+    """Return the minimum time from a list of time strings."""
+    time_list = Time(items)
+    return time_list.min()
+
+def maxdate(items):
+    """Return the maximum time from a list of time strings."""
+    time_list = Time(items)
+    return time_list.max()
+
+def mintime(items):
+    times = [_isotime(time_str) for time_str in items]
+    return min(times).isoformat()
+
+def maxtime(items):
+    times = [_isotime(time_str) for time_str in items]
+    return max(times).isoformat()
+
+def _isotime(time_str):
+    hms = [float(i) for i in time_str.split(':')]
+    sec_ms = hms[2] - int(hms[2])
+    isotime = time(int(hms[0]), int(hms[1]), int(hms[2]), int(sec_ms * 1000000))
+    return isotime
```

Comment:
That said the files that I have define DATE-END as `<DATE>T<TIME>` and `TIME-END` is not defined so the code will probably crash. However this needs to be fixed in the files and possibly in SDP.

Reply:
The _isotime() function is required in order to format the isoformat time strings so that the datetime library can understand them.  The astropy.time.Time class also does NOT know how to interpret the time strings (without a DATE) either:
In [104]: from astropy.time import Time                                                                                             

In [105]: t0 = '19:23:05.767'                                                                                                       

In [106]: at0 = Time(t0, format='isot', scale='utc')                                                                                
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-106-9be6aaa59336> in <module>
----> 1 at0 = Time(t0, format='isot', scale='utc')

/user/hack/dev/miniconda3/envs/rthstdev/lib/python3.6/site-packages/astropy/time/core.py in __init__(self, val, val2, format, scale, precision, in_subfmt, out_subfmt, location, copy)
    395         else:
    396             self._init_from_vals(val, val2, format, scale, copy,
--> 397                                  precision, in_subfmt, out_subfmt)
    398             self.SCALES = TIME_TYPES[self.scale]
    399 

/user/hack/dev/miniconda3/envs/rthstdev/lib/python3.6/site-packages/astropy/time/core.py in _init_from_vals(self, val, val2, format, scale, copy, precision, in_subfmt, out_subfmt)
    450         # Parse / convert input values into internal jd1, jd2 based on format
    451         self._time = self._get_time_fmt(val, val2, format, scale,
--> 452                                         precision, in_subfmt, out_subfmt)
    453         self._format = self._time.name
    454 

/user/hack/dev/miniconda3/envs/rthstdev/lib/python3.6/site-packages/astropy/time/core.py in _get_time_fmt(self, val, val2, format, scale, precision, in_subfmt, out_subfmt)
    501                 pass
    502         else:
--> 503             raise ValueError('Input values did not match {0}'.format(err_msg))
    504 
    505     @classmethod

ValueError: Input values did not match the format class isot

So, any keyword which has the time only string would need this function.  

The rules also can NOT support comparing a combination of multiple keywords, since the information for only 1 keyword is available at a time.  This prevents your suggestion of combining the DATE-END and TIME-END from being implemented.

Reply:
What about time series observations who are taken in different days?

Reply:
Obviously the two suggestions above go together. And comparing times only without the date does not make much sense. If the code doesn't support this then this issue probably does not have an easy fix.

Reply:
The data from TSO never actually get "combined", but rather concatenated/appended together within a single file. So theoretically we probably should then update these primary header keywords to reflect the total range of dates/times represented by the data, but for serious use of TSO time stamps folks will always use the data in the INT_TIMES table extension, not these generic header keywords.

Reply:
Are these rules for resampled data? TSO do not get resampled, right? But a dithered observation can span two dates and this will make the times wrong.

Reply:
The only way to avoid these type of problems is to force the observations to be sorted by expstart time when handed to model_blender.  You could then use 'first' and 'last' for the times, since it would pull the values from the first and last observation in the list and not scramble the times across different days. (for example, time-beg: 23:10:51.2, 00:15:23.5 would get incorrectly inverted just like Nadia was implying).  This is also the only way to blend on 2 (or more) keywords together like "<DATE-BEG>T<TIME-BEG>", again like Nadia was suggesting.

Reply:
Could that sorting be done within model_blender itself, before it does any of its computations on the meta values? There's no guarantee that users or the asn_generator will create ASN's with the members sorted in the appropriate order and I'd rather not have to stick some sorting function into each pipeline or step module that calls model_blender. It would be simpler to just have it in one place, i.e. in model_blender.